kv_storage.h

```diff
@@ -0,0 +1,383 @@
+/*******************************************************************************
+The MIT License (MIT)
+
+Copyright (c) 2015 Dmitry "Dima" Korolev <dmitry.korolev@gmail.com>
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*******************************************************************************/
+
+#ifndef SHERLOCK_KV_STORAGE_H
+#define SHERLOCK_KV_STORAGE_H
+
+#include <atomic>
+#include <future>
+#include <string>
+
+#include "sherlock.h"
+
+#include "../Bricks/mq/inmemory/mq.h"
+
+namespace sherlock {
+namespace kv_storage {
+// Base structures for user entry types.
+struct AllowNonThrowingGet {
+  constexpr static bool allow_nonthrowing_get = true;
+};
+
+enum NullEntryTypeHelper { NullEntry };
+struct Nullable {
+  const bool exists;
+  Nullable() : exists(true) {}
+  Nullable(NullEntryTypeHelper) : exists(false) {}
+};
+
+struct Deletable {};  // User promises to serialize Nullable::exists;
+
+// Universal null entry creation.
+template <typename T_ENTRY, bool IS_NULLABLE>
+struct GetNullEntry {
+  static T_ENTRY get() {
+    static_assert(true, "Should never happen.");
+    return T_ENTRY();
+  }
+};
+
+template <typename T_ENTRY>
+struct GetNullEntry<T_ENTRY, true> {
```

Comment:
I don't think `Get` is the best name here. `Create` or `Construct` is better. Also, since it's a class to help `CreateNullEntry()` below, why not simply `CreateNullEntryImpl`?